autograder/handgrading/tests/test_views/test_handgrading_result_views.py

```diff
@@ -670,71 +691,107 @@ def test_get_handgrading_results_exclude_staff(self):
-        if num_groups is None:
-            num_groups = num_results
+        """Perform the test for the handgrading results view."""
+        if submission_configs is None:
+            # default initialization of submission_configs to all finished grading submissions
+            submission_configs = [_GroupHGDataConfig()] * num_groups
         else:
-            assert num_groups >= num_results
+            assert len(submission_configs) == num_groups
+
+        # ensure that finished grading submissions are sorted before rejected submissions
+        submission_configs.sort(
+            key=lambda config:
+            config.submission_status is Submission.GradingStatus.finished_grading)
 
         groups = [obj_build.make_group(3, project=self.project) for _ in range(num_groups)]
-        expected_data = []
-        for i in range(num_results):
-            group = groups[i]
-            s = obj_build.make_finished_submission(group=group)
-            score = random.randint(0, self.rubric.max_points + 3)
-            hg_result = hg_models.HandgradingResult.objects.validate_and_create(
-                submission=s, group=group, handgrading_rubric=self.rubric,
-                points_adjustment=score,
-                finished_grading=bool(random.getrandbits(1)))
-
-            # Groups that have a handgrading result
-            data = hg_result.group.to_dict()
-            data['handgrading_result'] = {
-                'total_points': hg_result.total_points,
-                'total_points_possible': hg_result.total_points_possible,
-                'finished_grading': hg_result.finished_grading
-            }
-            data['member_names'].sort()
-            expected_data.append(data)
 
-        # Groups that don't have a handgrading result
-        for i in range(num_results, num_groups):
-            data = groups[i].to_dict()
-            data['member_names'].sort()
-            data['handgrading_result'] = None
-            expected_data.append(data)
+        def gen_exp_data(group: ag_models.Group, config: _GroupHGDataConfig):
+            """Generate the expected data for a group based on the configuration."""
+
+            submission = None
+            if config.submission_status is not None:
+                submission = obj_build.make_submission(
+                    group=group, status=config.submission_status)
+
+            if config.has_handgrading_result:
+                hg_result = hg_models.HandgradingResult.objects.validate_and_create(
+                    submission=submission,
+                    group=group, handgrading_rubric=self.rubric)
+                data = hg_result.group.to_dict()
+                data['handgrading_result'] = {
+                    'total_points': hg_result.total_points,
+                    'total_points_possible': hg_result.total_points_possible,
+                    'finished_grading': hg_result.finished_grading
+                }
+            else:
```

Comment:
Ah, so looking at what you have, I think this does cover all the conditions I listed, but the way different parts of of that logic are grouped together (some in the if else, some in the assignment to `data['has_handgradeable_submission']`) makes me have to think harder to make sure that mapping is there. Maybe that's a sign that we shouldn't be generating the expected data, but at present I choose to ignore that sign. ;)

To make the mapping between the code and the cases clearer, I'd request you add this condition:
```
elif not config.has_handgrading_result and config.submission_status == ag_models.Submission.GradingStatus.finished_grading:
    data = group.to_dict()
    data['handgrading_result'] = None
    data['has_handgradeable_submission'] = True
```
Similarly, let's explicitly assign has_handgradeable_submission in each branch.

I believe we do already have a test case for this: `test_get_paginated_results_some_groups_have_no_result`